### PR TITLE
[17.07] backport Fix requests for docker host ending with slash

### PR DESCRIPTION
**Pull request by andrewhsu**
backport fix:
* moby/moby/pull/34487 Fix requests for docker host ending with slash

with cherry pick of git commit https://github.com/moby/moby/pull/34487/commits/823e88d931 (no conflicts):
```
$ git cherry-pick -s -x -Xsubtree=components/engine 823e88d
```

and update of `components/cli/vendor/github.com/docker/docker` with the changes to `components/engine`

### CODE REVIEW

**Comment by andrewhsu** [comment]
Hmm...lots of changes in the vendor. I'll split that out to a new PR.

**Comment by andrewhsu** [comment]
I've backed out https://github.com/docker/docker-ce/pull/194/commits/cdf9130c398126e71187e248458f82328093d209 with a git reset.

File: components/cli/vendor/github.com/docker/docker/client/client.go [hunk 1]
```diff
@@ -260,8 +261,13 @@ func (cli *Client) NegotiateAPIVersionPing(p types.Ping) {
 		p.APIVersion = "1.24"
 	}
 
-	// if server version is lower than the current cli, downgrade
-	if versions.LessThan(p.APIVersion, cli.ClientVersion()) {
+	// if the client is not initialized with a version, start with the latest supported version
```
Comment (tonistiigi) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: This is not part of the upstream change. If it is needed then reference the upstream change from this PR.

Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: Looks like this was brought into the `components/engine` with git commit https://github.com/docker/docker-ce/commit/1a311235bab8cd31f8b3a63b03c34097f1d52818

Comment (andrewhsu) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: Which came from upstream commit https://github.com/moby/moby/commit/5975dc4b4b8a6fa2e0edd1f1b113e7f66dad732b

Comment (tonistiigi) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: Doesn't seem specifically needed for the cli use-case to me

Comment (tonistiigi) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: Actually, let me double check that

Comment (tonistiigi) [inline components/cli/vendor/github.com/docker/docker/client/client.go]: This does not look correct to me, so I would recommend removing to be safe. First, in the `cli` use case always sets the version, so the new condition doesn't apply. The new `versions.LessThan(p.APIVersion, api.DefaultVersion)` does not look correct to me. If user sets a manual low version it should not be made bigger if daemon supports something that is is lower than default and higher than the manual override. The cli does not seem to set the `client.manualOverride` as well. cc @thaJeztah

**Comment by andrewhsu** [comment]
@tonistiigi vendored code is now exact same diff as backported change to `components/engine`

**Comment by tonistiigi** [comment]
LGTM

**Comment by vieux** [comment]
LGTM

**Pull request merged**